doc/sphinx/proof-engine/ltac.rst

```diff
@@ -713,339 +962,563 @@ tactic
-   same types. This provides with a primitive form of matching under
-   context which does not require manipulating a functional term.
-
-   If the matching with :n:`@cpattern__1` succeeds, then :n:`@ltac_expr__1` is
-   evaluated into some value by substituting the pattern matching
-   instantiations to the metavariables. If :n:`@ltac_expr__1` evaluates to a
-   tactic and the match expression is in position to be applied to a goal
-   (e.g. it is not bound to a variable by a :n:`let in`), then this tactic is
-   applied. If the tactic succeeds, the list of resulting subgoals is the
-   result of the match expression. If :n:`@ltac_expr__1` does not evaluate to a
-   tactic or if the match expression is not in position to be applied to a
-   goal, then the result of the evaluation of :n:`@ltac_expr__1` is the result
-   of the match expression.
-
-   If the matching with :n:`@cpattern__1` fails, or if it succeeds but the
-   evaluation of :n:`@ltac_expr__1` fails, or if the evaluation of
-   :n:`@ltac_expr__1` succeeds but returns a tactic in execution position whose
-   execution fails, then :n:`cpattern__2` is used and so on. The pattern
-   :n:`_` matches any term and shadows all remaining patterns if any. If all
-   clauses fail (in particular, there is no pattern :n:`_`) then a
-   no-matching-clause error is raised.
-
-   Failures in subsequent tactics do not cause backtracking to select new
-   branches or inside the right-hand side of the selected branch even if it
-   has backtracking points.
+   .. todo 30 May 20: the `@?ident` form is in dangling_pattern_extension_rule, not included in the doc yet
+      maybe belongs with "Applications"
+
+   `lazymatch`
+      Causes the match to commit to the first matching branch
+      rather than trying a new match if :n:`@ltac_expr` fails.
+      :ref:`Example<match_vs_lazymatch_ex>`.
+
+   `match`
+      If :n:`@ltac_expr` fails, continue matching with the next branch.
+      Failures in subsequent tactics (after the `match`) will not cause selection
+      of a new branch.  Examples :ref:`here<match_vs_lazymatch_ex>` and
+      :ref:`here<match_vs_multimatch_ex>`.
+
+   `multimatch`
+      If :n:`@ltac_expr` fails, continue matching with the next branch.
+      When an :n:`@ltac_expr` succeeds for a branch, subsequent failures
+      (after the `multimatch`) causing consumption of all the backtracking points
+      (if any) in :n:`@ltac_expr` triggers selection of a new matching branch.
+      :ref:`Example<match_vs_multimatch_ex>`.
+
+      :tacn:`match` :n:`…` is, in fact, shorthand for :tacn:`once` :tacn:`multimatch` `…`.
+
+   :n:`@cpattern`
+      The syntax of :token:`cpattern` is
+      the same as that of :token:`term`\s, but it can contain pattern matching
+      metavariables in the form :n:`?@ident`.  :g:`_` can be used to match
+      irrelevant terms.  :ref:`Example<match_with_holes_ex>`.
+
+      .. todo PR Didn't understand the following paragraph well enough to revise
```

Comment:
Here is an example (perhaps not the best one, because it relies on the trick of `match a with b => c end` reducing the `match` away):
```coq
Ltac uncurry f :=
  lazymatch f with
  | fun (a : ?A) (b : ?B) => ?f (* note the lack of `@?` *)
    => let ab := fresh in
       uncurry (fun ab : A * B => match fst ab, snd ab with a, b => f end)
  | ?f => f
  end.

Goal True.
  let x := uncurry (fun (a b c : unit) (d e : nat) => (a, b, c, d, e)) in
  pose x.
  (* 1 subgoal (ID 38)

  p := fun H0 : unit * unit * unit * nat * nat =>
       (fst (fst (fst (fst H0))), snd (fst (fst (fst H0))),
       snd (fst (fst H0)), snd (fst H0), snd H0)
   : unit * unit * unit * nat * nat -> unit * unit * unit * nat * nat
  ============================
  True
   *)
```
Here is a variant that doesn't use that
```coq
Ltac uncurry f :=
  lazymatch (eval cbv beta in f) with
  | fun (a : ?A) (b : ?B) => ?f (* note the lack of `@?` *)
    => let ab := fresh in
       uncurry ((fun F (ab : A * B) => F (fst ab) (snd ab)) (fun (a : A) (b : B) => f))
  | ?f => f
  end.

Goal True.
  let x := uncurry (fun (a b c : unit) (d e : nat) => (a, b, c, d, e)) in
  pose x.
  (* 1 subgoal (ID 38)

  p := fun H0 : unit * unit * unit * nat * nat =>
       (fst (fst (fst (fst H0))), snd (fst (fst (fst H0))),
       snd (fst (fst H0)), snd (fst H0), snd H0)
   : unit * unit * unit * nat * nat -> unit * unit * unit * nat * nat
  ============================
  True
   *)
```
Does this explain it?

Reply:
Note also that you are already making use of this feature in `_match_with_holes_ex:`